pkg/services/ngalert/eval/eval.go

```diff
@@ -67,6 +70,28 @@ func (r *conditionEvaluator) EvaluateRaw(ctx context.Context, now time.Time) (re
 		}
 	}()
 
+	// patching the hysteresis commands with the current loaded metrics.
+	// This optimization is to save resources on marshalling AlertQuery.
+	// Mutex is to make sure that the evaluation is synchronous.
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	if r.loadedMetricsReader != nil {
+		cmds := expr.GetCommandsFromPipeline[*expr.HysteresisCommand](r.pipeline)
+		if len(cmds) > 0 {
+			loaded := r.loadedMetricsReader.Read()
+			logger.FromContext(ctx).Debug("Populating loaded metrics for hysteresis command", "commands", len(cmds), "loaded", len(loaded))
+			for _, cmd := range cmds {
+				cmd.LoadedDimensions = loaded
+			}
+			// reset the data after evaluation
+			defer func() {
+				for _, cmd := range cmds {
+					cmd.LoadedDimensions = nil
+				}
+			}()
+		}
+	}
```

Comment:
During the review session the request was to replace this code with one that adds `loadedMetrics` data frame into the query model during conversion from AlertQuery to server-side expressions Request. This will let us reuse the code in the cloud alerting ruler.